### PR TITLE
Add unit tests

### PR DESCRIPTION
**Pull request by MrMeison**
- Update API to 1.5.0
- Add vscode settings
- Add travis and coveralls integrations
- Fix empty children bug
- Migrate from typings to npm types repository

### CODE REVIEW

File: src/visual.ts [hunk 1]
```diff
@@ -246,7 +245,9 @@ module powerbi.extensibility.visual {
             if (!options
                 || !options.dataViews
                 || !options.dataViews[0]
-                || !options.dataViews[0].matrix) {
+                || !options.dataViews[0].matrix
+                || !options.dataViews[0].matrix.rows.root.children
+                || !options.dataViews[0].matrix.rows.root.children.length) {
```
Comment (ignatvilesov) [inline src/visual.ts]: Could you please add ```options.dataViews[0].matrix.rows``` to this condition too

File: test/_references.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ *  THE SOFTWARE.
+ */
+
+// External
+/// <reference path="../node_modules/@types/jasmine/index.d.ts" />
+/// <reference path="../node_modules/@types/jasmine-jquery/index.d.ts" />
```
Comment (ignatvilesov) [inline test/_references.ts]: Is it necessary? As far as I know, TypeScript finds types in ```node_modules/@types``` automatically.

File: test/visualData.ts [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    // powerbi.extensibility.utils.test
+    import getRandomNumber = powerbi.extensibility.utils.test.helpers.getRandomNumber;
+    import testData = powerbi.extensibility.visual.test.data;
+    import TestDataViewBuilder = powerbi.extensibility.utils.test.dataViewBuilder.TestDataViewBuilder;
+    import TestDataViewBuilderCategoryColumnOptions = powerbi.extensibility.utils.test.dataViewBuilder.TestDataViewBuilderCategoryColumnOptions;
+
+    interface INamed {
+        name: string;
+    }
+    export class SunburstData extends TestDataViewBuilder {
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add an empty line after declaration of the interface.

File: test/visualData.ts [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+        public readonly RegionsDataSet = "Regions";
+        public readonly CountriesDataSet = "Countries";
+        public readonly StatesDataSet = "States";
+        public readonly visualName: string = "Sunburst";
+        private readonly measure: powerbi.DataViewMetadataColumn = {
+            displayName: "Value",
+            index: 3,
+            isMeasure: true,
+            queryName: "Sum(" + this.visualName + ".Value)",
```
Comment (ignatvilesov) [inline test/visualData.ts]: Could you please use ES6 template strings?

File: test/visualData.ts [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+    interface INamed {
+        name: string;
+    }
+    export class SunburstData extends TestDataViewBuilder {
+        public readonly RegionsDataSet = "Regions";
+        public readonly CountriesDataSet = "Countries";
+        public readonly StatesDataSet = "States";
+        public readonly visualName: string = "Sunburst";
+        private readonly measure: powerbi.DataViewMetadataColumn = {
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please remove ```powerbi.``` here.

File: test/visualData.ts [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+            queryName: "Sum(" + this.visualName + ".Value)",
+            roles: { Values: true }
+        };
+
+        private countElements: number = 3;
+        public set countGeneratedElements(count: number) {
+            this.countElements = count;
+        }
+        getDataView(columnNames?: string[]): DataView {
+            if (!columnNames) return;
```
Comment (ignatvilesov) [inline test/visualData.ts]: Could you please add curly brackets around the return statement?

File: test/visualData.ts [hunk 7]
```diff
@@ -0,0 +1,148 @@
...
+            let testData: INamed[][] = [];
+            let columns: DataViewMetadataColumn[] = [];
+            columnNames.forEach((columnName, index) => {
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type for each argument of the arrow function.

File: test/visualData.ts [hunk 8]
```diff
@@ -0,0 +1,148 @@
...
+
+        private readonly allData: { [name: string]: INamed[] } =
+        {
+            "Regions": testData.Regions,
+            "Countries": testData.Countries,
+            "States": testData.States
+        };
+
+        private buildMatrix(data: INamed[][], columns: DataViewMetadataColumn[], measure: DataViewMetadataColumn): DataViewMatrix {
+            let rootNode = this.generateNode(null, data);
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 9]
```diff
@@ -0,0 +1,148 @@
...
+            let matrix = {
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+                    }]
+                },
+                valueSources: [this.measure]
+            };
+            return matrix;
+        }
+
+        private generateColumn(columnName: string, index: number, visualName?: string): DataViewMetadataColumn {
+            visualName = visualName || "testVisual";
+            return { displayName: columnName, index: index, queryName: visualName + "." + columnName, roles: { Values: true } };
```
Comment (ignatvilesov) [inline test/visualData.ts]: Could you please use a template string instead of ```visualName + "." + columnName```.

File: test/visualData.ts [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        private generateNode(head: INamed, data: INamed[][], level: number = -1): DataViewMatrixNode {
+            let treeNode: DataViewMatrixNode;
+            let nextLevel = level + 1;
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 12]
```diff
@@ -0,0 +1,148 @@
...
+            if (level === -1) {
+                treeNode = {
+                    children: []
+                };
+            } else {
+                let isLastLevel = !data[nextLevel];
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+            }
+            if (data[nextLevel]) {
+                data[nextLevel].forEach((item) => {
+                    treeNode.children.push(this.generateNode(item, data, nextLevel));
+                });
+            }
+            return treeNode;
+        }
+        private getRandomArrayElements(arr: any[], count: number): any[] {
+            if (arr.length < count) return arr;
```
Comment (ignatvilesov) [inline test/visualData.ts]: Could you please add curly brackets around the return statement.

File: test/visualData.ts [hunk 14]
```diff
@@ -0,0 +1,148 @@
...
+            let shuffled = arr.slice(0),
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+                i = arr.length,
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 16]
```diff
@@ -0,0 +1,148 @@
...
+                min = i - count,
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+                temp,
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+                index;
```
Comment (ignatvilesov) [inline test/visualData.ts]: Please add a type of the variable.

File: test/visualData.ts [hunk 19]
```diff
@@ -0,0 +1,148 @@
...
+            while (i-- > min) {
+                index = Math.floor((i + 1) * Math.random());
+                temp = shuffled[index];
+                shuffled[index] = shuffled[i];
+                shuffled[i] = temp;
+                shuffled[i] = temp;
+                shuffled[i] = temp;
```
Comment (ignatvilesov) [inline test/visualData.ts]: The ```shuffled[i] = temp;``` repeats 3 times. Please remove unnecessary lines.

File: test/visualTest.ts [hunk 20]
```diff
@@ -0,0 +1,156 @@
...
+
+/// <reference path="_references.ts"/>
+
+namespace powerbi.extensibility.visual.test {
+    // powerbi.extensibility.visual.test
+    import VisualData = powerbi.extensibility.visual.test.SunburstData;
+    import VisualBuilder = powerbi.extensibility.visual.test.SunburstBuilder;
+    import colorHelpers = powerbi.extensibility.utils.test.helpers.color;
+
+
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Please remove this empty line.

File: test/visualTest.ts [hunk 21]
```diff
@@ -0,0 +1,156 @@
...
+    import DataView = powerbi.DataView;
+    import DataViewMetadata = powerbi.DataViewMetadata;
+    import createSelectionId = powerbi.extensibility.utils.test.mocks.createSelectionId;
+
+    // Sunburst1445472000808
+    import VisualClass = powerbi.extensibility.visual.Sunburst1445472000808.Sunburst;
+    import SunburstSlice = powerbi.extensibility.visual.Sunburst1445472000808.SunburstSlice;
+
+    const DEFAULT_WAIT_FOR_RENDER = 500;
+    const VISUAL_SELECTOR = ".sunBurstDrawArea";
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Could you please use pascal case for static variables and consts.

File: test/visualTest.ts [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+    import DataViewMatrix = powerbi.DataViewMatrix;
+    import DataView = powerbi.DataView;
+    import DataViewMetadata = powerbi.DataViewMetadata;
+    import createSelectionId = powerbi.extensibility.utils.test.mocks.createSelectionId;
+
+    // Sunburst1445472000808
+    import VisualClass = powerbi.extensibility.visual.Sunburst1445472000808.Sunburst;
+    import SunburstSlice = powerbi.extensibility.visual.Sunburst1445472000808.SunburstSlice;
+
+    const DEFAULT_WAIT_FOR_RENDER = 500;
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Could you please use pascal case for static variables and consts.

File: test/visualTest.ts [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+                visualBuilder.update(dataView);
+                visualBuilder.updateRenderTimeout(updatedDataView, () => {
+                    expect($(`${VISUAL_SELECTOR} path`).length).toBe(40);
+                    done();
+                }, DEFAULT_WAIT_FOR_RENDER);
+            }, DEFAULT_WAIT_FOR_RENDER);
+        });
+
+        describe("Labels", () => {
+
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Please remove this empty line.

File: test/visualTest.ts [hunk 24]
```diff
@@ -0,0 +1,156 @@
...
+            defaultDataViewBuilder: VisualData,
+            dataView: DataView;
+
+        beforeEach(() => {
+            visualBuilder = new VisualBuilder(500, 500);
+            defaultDataViewBuilder = new VisualData();
+        });
+
+        it('sunburst slices dom validation', (done) => {
+
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Please remove this empty line.

File: test/visualTest.ts [hunk 25]
```diff
@@ -0,0 +1,156 @@
...
+            it('category labels should be visible by default', (done) => {
+                let dataView: DataView = defaultDataViewBuilder.getDataView(
+                    [
+                        defaultDataViewBuilder.RegionsDataSet,
+                        defaultDataViewBuilder.CountriesDataSet,
+                        defaultDataViewBuilder.StatesDataSet
+                    ]);
+
+                visualBuilder.updateRenderTimeout(dataView, () => {
+                    const firstPoint = visualBuilder.mainElement.find("path").last();
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Could you please add a type of the variable.

File: test/visualTest.ts [hunk 26]
```diff
@@ -0,0 +1,156 @@
...
+                        defaultDataViewBuilder.RegionsDataSet,
+                        defaultDataViewBuilder.CountriesDataSet
+                    ]);
+
+                dataView.metadata.objects = {
+                    group: { showSelected: false }
+                };
+
+                visualBuilder.updateRenderTimeout(dataView, () => {
+                    const firstPoint = visualBuilder.mainElement.find("path").last();
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Could you please add a type of the variable.

File: test/visualTest.ts [hunk 27]
```diff
@@ -0,0 +1,156 @@
...
+                    firstPoint.d3MouseDown(5, 5);
+                    setTimeout(() => {
+                        expect($(`${VISUAL_SELECTOR} text[style*='opacity: 1']`).length).toBe(1);
+                        done();
+                    }, DEFAULT_WAIT_FOR_RENDER);
+                }, DEFAULT_WAIT_FOR_RENDER);
+            });
+        });
+        describe("Test invalid input data", () => {
+
```
Comment (ignatvilesov) [inline test/visualTest.ts]: Please remove this empty line.

**Comment by MrMeison** [comment]
PR was updated. Please, check it again.

File: src/visual.ts [hunk 28]
```diff
@@ -246,7 +245,10 @@ module powerbi.extensibility.visual {
             if (!options
                 || !options.dataViews
                 || !options.dataViews[0]
-                || !options.dataViews[0].matrix) {
+                || !options.dataViews[0].matrix
+                || !options.dataViews[0].matrix.rows
+                || !options.dataViews[0].matrix.rows.root.children
+                || !options.dataViews[0].matrix.rows.root.children.length) {
```
Comment (ignatvilesov) [inline src/visual.ts]: Could you please include ```options.dataViews[0].matrix.rows.root``` to the this condition above.

File: src/visual.ts [hunk 29]
```diff
@@ -296,7 +298,7 @@ module powerbi.extensibility.visual {
                     }
 
                     // TODO: why do we need to do this lie below?
-                    d3.selectAll("path") // TODO: we shouldn't find all paths in the DOM.
+                    d3.selectAll("path") // TODO: we shouldn"t find all paths in the DOM.
```
Comment (ignatvilesov) [inline src/visual.ts]: I suppose that you must use a single quote for ```shouldn"t```.

File: package.json [hunk 30]
```diff
@@ -1,31 +1,51 @@
...
+    "karma-chrome-launcher": "2.0.0",
+    "karma-coverage": "1.1.1",
+    "karma-jasmine": "1.1.0",
+    "karma-remap-istanbul": "0.6.0",
+    "karma-sourcemap-loader": "0.3.7",
+    "karma-typescript-preprocessor": "0.3.1",
+    "powerbi-visuals-tools": "1.5.0",
+    "powerbi-visuals-utils-testutils": "0.2.2",
+    "tslint": "4.5.1",
+    "typescript": "2.1.4"
```
Comment (AviSander) [inline package.json]: Please also add:
"tslint-microsoft-contrib": "^4.0.0"

File: package.json [hunk 31]
```diff
@@ -1,31 +1,51 @@
...
   },
   "scripts": {
-    "postinstall": "typings install && pbiviz update 1.3.0",
-    "typings": "typings",
+    "postinstall": "pbiviz update 1.5.0",
     "pbiviz": "pbiviz",
     "start": "pbiviz start",
     "package": "pbiviz package",
-    "lint": "node node_modules/tslint/bin/tslint \"+(src|test)/**/*.ts\""
+    "lint": "tslint \"+(src|test)/**/*.ts\"",
```
Comment (AviSander) [inline package.json]: please modify to:
 "lint": "node node_modules/tslint/bin/tslint -r \"node_modules/tslint-microsoft-contrib\"  \"+(src|test)/**/*.ts\"",

**Pull request merged**